pkg/apis/configuration/v1alpha1/kong_custom_entity_types.go

```diff
@@ -0,0 +1,94 @@
+	KongCustomEntityKind = "KongCustomEntity"
+)
+
+type KongEntityScope string
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Namespaced,shortName=kce,categories=kong-ingress-controller,path=kongcustomentities,singular=kongcustomentity
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Entity Type",type=string,JSONPath=`.type`,description="type of the Kong entity"
+// +kubebuilder:validation:XValidation:rule="self.scope == oldSelf.scope",message="The scope field is immutable"
+// +kubebuilder:validation:XValidation:rule="self.type == oldSelf.type",message="The type field is immutable"
+// +kubebuilder:validation:XValidation:rule="!(self.type in ['services','routes','upstreams','targets','plugins','consumers','consumer_groups'])",message="The type field cannot be known Kong entity types"
+// +kubebuilder:validation:XValidation:rule="!(self.scope == 'independent' && has(self.parentRef)) && !(self.scope == 'attached' && !has(self.parentRef))",message="attached KongCustomEntity must have parentRef; independent KongCustomEntity must not have parentRef"
+// REVIEW: put all fields other than "status" under the "spec"?
+
+// KongCustomEntity defines a "custom" Kong entity that KIC cannot support the entity type directly.
+type KongCustomEntity struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Enum=independent;attached
+	// REVIEW: We can get the schema of entities from Kong gateway admin APIs including fields requiring foreign reference.
+	// Is the "Scope" field really required?
+
+	// Scope means whether the entity can be specified independently.
+	// "independent" means that the entity does not depend on other entities;
+	// "attached" means that the entity has "foreign" fields referring to other entities.
+	Scope KongEntityScope `json:"scope"`
+
+	// EntityType is the type of the Kong entity. The type is used in generating declarative configuration.
+	EntityType string `json:"type"`
+	// Fields defines the fields of the Kong entity itself.
+	Fields apiextensionsv1.JSON `json:"fields"`
+	// ControllerName specifies the controller that should reconcile it, like ingress class.
+	ControllerName string `json:"controllerName"`
+
+	// ParentRef references the kubernetes resource it attached to when its scope is "attached".
+	// Currently only KongPlugin/KongClusterPlugin allowed. This will make the custom entity to be attached
+	// to the entity(service/route/consumer) where the plugin is attached.
+	ParentRef *ObjectReference `json:"parentRef,omitempty"`
+	// Status stores the reconciling status of the resource.
+	Status KongCustomEntityStatus `json:"status,omitempty"`
+}
+
+// REVIEW:
+// - Should we define dedicated type aliases for each field (like gateway API does, define types "Group","Kind","Namespace","ObjectName")?
```

Comment:
More @shaneutt questions: what was the motivation for aliasing these in the GWAPI case? Does that same motivation apply here?

Reply:
After some discussion, I think the purpose in GWAPI could be:
- Add validation expressions on these dedicated types so that the validation could be shared by different types that includes these fields
- Prevent from assigning values from a field to another field (like `a.Kind = b.Group`, if we define both as `string` it can pass compilation, but if `Group` and `Kind` have their own types, the mistake can be checked by compiler).
Since the type is expected to be only used in `KongCustomEntity` CRD but not publicly used like types in GWAPI, I decide not to define dedicated types.